sdk/runner/sdk.go

```diff
@@ -0,0 +1,192 @@
+package runnersdk
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	cosmostypes "github.com/cosmos/cosmos-sdk/types"
+	"github.com/mesg-foundation/engine/container"
+	"github.com/mesg-foundation/engine/cosmos"
+	"github.com/mesg-foundation/engine/hash"
+	"github.com/mesg-foundation/engine/instance"
+	"github.com/mesg-foundation/engine/protobuf/api"
+	"github.com/mesg-foundation/engine/runner"
+	accountsdk "github.com/mesg-foundation/engine/sdk/account"
+	instancesdk "github.com/mesg-foundation/engine/sdk/instance"
+	servicesdk "github.com/mesg-foundation/engine/sdk/service"
+	"github.com/mesg-foundation/engine/x/xos"
+	"github.com/tendermint/tendermint/mempool"
+)
+
+// SDK is the runner sdk.
+type SDK struct {
+	cdc          *codec.Codec
+	accountSDK   *accountsdk.SDK
+	serviceSDK   servicesdk.Service
+	instanceSDK  *instancesdk.SDK
+	client       *cosmos.Client
+	container    container.Container
+	port         string
+	engineName   string
+	ipfsEndpoint string
+}
+
+// New returns the runner sdk.
+func New(cdc *codec.Codec, client *cosmos.Client, accountSDK *accountsdk.SDK, serviceSDK servicesdk.Service, instanceSDK *instancesdk.SDK, container container.Container, engineName, port, ipfsEndpoint string) *SDK {
+	sdk := &SDK{
+		cdc:          cdc,
+		container:    container,
+		accountSDK:   accountSDK,
+		serviceSDK:   serviceSDK,
+		instanceSDK:  instanceSDK,
+		client:       client,
+		port:         port,
+		engineName:   engineName,
+		ipfsEndpoint: ipfsEndpoint,
+	}
+	return sdk
+}
+
+// Create creates a new runner.
+func (s *SDK) Create(req *api.CreateRunnerRequest, accountName, accountPassword string) (*runner.Runner, error) {
+	account, err := s.accountSDK.Get(accountName)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: pass account by parameters
+	accNumber, accSeq := uint64(0), uint64(0)
+	user, err := cosmostypes.AccAddressFromBech32(account.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	// calculate instance's hash.
+	// TODO: this should be merged with the same logic currently in instance sdk
+	srv, err := s.serviceSDK.Get(req.ServiceHash)
+	if err != nil {
+		return nil, err
+	}
+	instanceEnv := xos.EnvMergeSlices(srv.Configuration.Env, req.Env)
+	envHash := hash.Dump(instanceEnv)
+	// TODO: should be done by instance
+	instanceHash := hash.Dump(&instance.Instance{
+		ServiceHash: srv.Hash,
+		EnvHash:     envHash,
+	})
+
+	// start the container
+	imageHash, err := build(s.container, srv, s.ipfsEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	_, err = start(s.container, srv, instanceHash, imageHash, instanceEnv, s.engineName, s.port)
+	if err != nil {
+		return nil, err
+	}
+	onError := func() {
+		stop(s.container, instanceHash, srv.Dependencies)
+	}
+
+	msg := newMsgCreateRunner(s.cdc, user, req.ServiceHash, envHash)
+	tx, err := s.client.BuildAndBroadcastMsg(msg, accountName, accountPassword, accNumber, accSeq)
+	if err != nil {
+		defer onError()
+		if err == mempool.ErrTxInCache {
```

Comment:
this shouldn't be necessary with https://github.com/mesg-foundation/engine/pull/1462

Reply:
I think the issue still exists, because of two clients might get the same seq number.